### ISSUE TITLE
Jev型 decision model として盤面→操作確率を学習する

### ISSUE THREAD

**Issue by bottlenome** [bottlenome/Moduled2x2x2Cube#1]
## 背景
従来研究では 2×2×2 Rubik's Cube の盤面からゴールまでの距離 d(s) を予測していた。

Jev のような non-autoregressive decision model の考え方をマップすると、距離予測（value）ではなく、盤面から次の操作の確率分布（policy）を直接出力する構成が考えられる。

## アイデア

固定 action head:
```
s = board state
p(a|s) = softmax(W Encoder(s))
```

候補操作を入力に含める Jev-like な構成:
```
score(s, a_i) = Model(s, a_i)
p(a_i|s) = softmax_i(score(s, a_i))
```

後者では action を固定クラスとして重みに埋め込まず、state と candidate action の関係を学習する。

## 検討事項
- 既存の距離予測 d(s) と policy P(a|s) の性能・学習効率比較
- 教師を「最短経路上の次手」とした場合、複数の最適手をどう分布として与えるか
- 固定 action head と (state, action) scoring の比較
- calibration（出力確率が実際の最適手率と対応するか）
- 距離別の generalization
- action candidate を変えた場合の generalization

## ロードマップ：2×2×2 Cube の最短距離 solver まで

最終目標は、任意の盤面から **最短手数で solved state に到達する NN-based solver** を構築すること。まず exact distance table を教師として使い、Jev 型 policy が最短経路を再現できるかを見る。

### Phase 0: Exact oracle / dataset
- 2×2×2 の状態表現と全合法手を確定
- solved state から BFS して全状態の最短距離 `d(s)` を計算
- 各状態について全操作 `a` の遷移 `T(s,a)` を生成
- `d(T(s,a)) = d(s)-1` となる操作集合 `A*(s)` を exact optimal actions とする
- train / validation / test を距離別にも評価できる形で固定

この exact table は学習対象ではなく、教師生成と最短性の検証 oracle として扱う。

### Phase 1: 既存 distance model baseline
既存方式:

```
s -> model -> d_hat(s)
```

各候補手について次状態を評価し、

```
a = argmin_a d_hat(T(s,a))
```

で greedy solve。

評価:
- distance prediction accuracy
- solve rate
- optimal solve rate
- optimal path に対する excess moves

### Phase 2: 固定 action-head policy
盤面から直接、次の操作確率を出す。

```
s -> Encoder -> logits(actions) -> P(a|s)
```

教師分布は

```
P*(a|s) = 1 / |A*(s)|   if a in A*(s)
          0              otherwise
```

として、複数の最短手をすべて正解として扱う。

まず greedy:

```
a_t = argmax_a P(a|s_t)
```

だけで solved state まで rollout する。

### Phase 3: Jev-like state/action scoring
action を固定 class head にせず、候補としてモデルに入力する。

```
score(s, a_i) = Model(s, a_i)
P(a_i|s) = softmax_i(score(s, a_i))
```

Phase 2 と同じ教師を使い、

- fixed action classification
- state/action relational scoring

を直接比較する。

### Phase 4: 最短距離 solver
policy の greedy rollout だけで最短性が保証されない場合、policy を探索 ordering として使う。

```
frontier = [start]

while frontier:
    s = pop(frontier)
    if solved(s):
        return path

    actions = sort_by_probability(P(a|s))
    expand(actions)
```

BFS / A* / beam search と組み合わせ、

1. solution を必ず発見できる
2. solution length = exact distance `d(s_start)`

を満たす solver にする。

特に「NN単体の greedy policy だけでどこまで exact shortest path を再現できるか」と「探索を加えた場合にどこまで探索量を削減できるか」を分離して測る。

### Phase 5: 最終評価
全 test state / 距離別に以下を測定。

- next-action optimality: `argmax P(a|s) in A*(s)`
- solve rate
- exact shortest solve rate
- solution length - optimal distance
- explored states / expanded nodes
- inference 回数
- calibration: confidence と optimal-action accuracy の対応

比較対象:

```
Exact BFS oracle
Distance model + greedy/search
Fixed-head policy
Jev-like (state, action) policy
Jev-like policy + search
```

### 到達条件
最終的には、盤面 `s` を入力すると

```
solve(s) -> [a1, a2, ..., ak]
```

を返し、

```
k = d(s)
```

となることを exact oracle で検証する。

その上で、**距離を明示的に予測する value 型と、次の操作分布を直接学習する decision/policy 型のどちらが、有限状態空間の最短経路構造を効率よく学習できるか**を研究上の比較軸とする。

---
Moved to bottlenome/generic_cube: https://github.com/bottlenome/generic_cube/issues/1
